### ISSUE TITLE
Download Pools?

### ISSUE THREAD

**Issue by huskelord** [lurkbbs/e621dl#8]
Just a thought but would there be a way to download pools of images and have them be downloaded under in a folder of the name of the pool?
For example if i had a file for artist A and artist A had made a comic  could I get a sub folder that would be named after that comic inside of a subfolder that has other comics by that author

Also please finsih the other issue first this was just a thought that come into my head that i thought i would ask.
Edit1: edited the wording after i came back to it since it was a little wierd.

**Comment by lurkbbs** [comment]
I thought about it myself. Unfortunately, there is no api for that, not even as simple as "is this pic a part of some pool". It can be done by first acquiring all pools' info to some database and checking from it and downloading it all, also with ignoring prefilter and section filter. Not in this week, maybe at the end of the month.

And yes, only after that hanging issue

**Comment by huskelord** [comment]
I came up with an idea that may help a bit with the download pool when you come back to it. 
basicley have the each of the dowloaded pools be made into a subfolder named after what ever the pool is called all in a subfolder called downloaed poos in the main folder.

basicly this

this should help keep everyting in order in the downloads

Downloads
|
|--Artist 1 (main folder)
|   |
|   |-- Downloaded Pools (subfolder)
|   |    |
|   |    |-- "insert Pool name"
|   |
|   |-- Artist 1 images (subfolder)
|   |    |
|   |    |-- images
|   |
|   |-- Artist 1 animations (subfolder)
|       |
|       |-- animations
|
|--Artist 2 (main folder)
|   |
|   |-- Downloaded Pools (subfolder)
|   |    |
|   |    |-- "insert Pool name" 
|   |
|   |-- Artist 2 images (subfolder)
|   |    |
|   |    |-- images
|   |
|   |-- Artist 2 animations (subfolder)
|       |
|       |-- animations

Edit1: changed up some things like I did above to make it easier to read.

**Comment by lurkbbs** [comment]
Yup, that's how I see it.

**Comment by huskelord** [comment]
this is somethig that i was wondering if would be possible but would there be a way to blacklist anyting that would be in a pool that way say i have a pool seperated for this artist would it be possible to make shure it doesn't show up anywhere else except in the pool folder like

Downloads
|
|--Artist 1 (main folder)
|   |
|   |-- Downloaded Pools (subfolder)
|   |    |
|   |    |-- "insert Pool name"
|   |
|   |-- Artist 1 images (subfolder)
|   |    |
|   |    |-- images -images from "insert Pool name"
|   |
|   |-- Artist 1 animations (subfolder)
|       |
|       |-- animations -animations from "insert Pool name"
|

once again i know this is going to be alot of work to do once you get to it so please dont worry about it until your ready and everyting else has been taken care of.

**Comment by lurkbbs** [comment]
Yep, it is possible, but really not soon.

**Comment by huskelord** [comment]
Figured as much. No worries take your time.

**Comment by lurkbbs** [comment]
OK, more than half a year have passed. I did nothing I promised. Looks like I won't do it in the near future. Even API of the site had changed, now every post has info about pools it's in. So, another thing I'll definitely do. I'll add simple default option that will autogenerate pools-grabbing config and donwload it after every other configs are done.

One month. I promise, by the 2020.06.08 it will be done.

**Comment by huskelord** [comment]
got it. I've been gone awhile myself so no worries.

**Comment by lurkbbs** [comment]
So, here it is, at last.
https://github.com/lurkbbs/e621dl/releases/tag/v5.11.0

Limitations:

- generated config for pools is one and the same, no way to make different settings for different kind of pools. Probably solvable
- pools identified only by their ids, no named folders. That is too hard to implement in this app. I can probably make a helper app for cases like that, but not much else, end even then they would be need to run separately.

Tell me when you read this and close it when you found there are no bugs.

**Comment by huskelord** [comment]
Having it run now

**Comment by huskelord** [comment]
Sorry for any confusion caused by what i put above

**Comment by huskelord** [comment]
Deleted my earlier commets since they were gibberish

Ran the program but had some problems with the pools going into the default subfolders

can you make it so the  _the pool_post_strategy =_ line is necessary in every section of code you want it implemented

this should be similar to how the _blacklist_default_subfolders =_ line is necessary for areas of code the default subfolders are not wanted

**Comment by lurkbbs** [comment]
Just to make sure, you need a way to disable pools for specific folders while enabling them for all other in `Defaults`, right?

I guess I can add `pool_post_strategy = none` option for that.

**Comment by huskelord** [comment]
other way around 

I need certain folders to be manualy enabled for pools with all others inculding the subfodlers of the enabled folders disabled

As it is now any main folder i put the pools _ the pool_post_strategy =_ line into will have pools copied in each of its subfolders which i do not want.

With my first config like this 

[Defaults]
days = 365000000
ratings = q, e
min_score = 0
min_favs = 0
max_downloads = inf
subfolders = .Images .Animated
posts_from = api

[Artist/Artist 1]
Tags = Artist 1
pool_post_strategy = move

[Artist/Artist 2]
Tags = Artist 2

[Artist/Artist 3]
Tags = Artist 3
blacklist_default_subfolders = true

I want to get like below in my downloads folder

Downloads
|
|--Artist 1 (main folder)
| |
| |-- Downloaded Pools (subfolder)
| | |
| | |-- "insert Pool name"
| | |
| | |-- etc.
| |
| |-- .Images (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.
| |
| |-- .Animations (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.

Downloads
|
|--Artist 2 (main folder)
| |
| |-- .Images (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.
| |
| |-- .Animations (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.

Downloads
|
|--Artist 3 (main folder)
| |
| |-- All of Artist 3 posts

**Comment by lurkbbs** [comment]
If it's not like that, it's a bug. Did you tried it already?

**Comment by huskelord** [comment]
yep thats why i brought it up.

**Comment by lurkbbs** [comment]
OK, this weekend tops, probably in about an hour or two.

**Comment by lurkbbs** [comment]
Ah. I guess I didn't understood the question well.

Yes, pools are created in subfolders. And yes, it looks like there is no way to disable it in this version. The way it's done, there is no simple (means in two hours) way to make it work. Best bet for now is multiple configs, one with pools but without subfolders, and another is the opposite. I'll try to think of something this weekend. I'll add "no_pools_in_subfolders" option.

Also, pool names instead of theirs id are not really possible in a single app, not without heavy rewrite. 
I think I should elaborate more. See, everything I have for a pic is this:

```ini
{
			"id": 2233791,
			"created_at": "2020-04-28T00:21:19.480-04:00",
			"updated_at": "2020-07-07T06:58:53.121-04:00",
			"file": {
				"width": 732,
				"height": 1131,
				"ext": "png",
				"size": 373012,
				"md5": "a8e539d5a9e2098875d9861287d04095",
				"url": "https://static1.e621.net/data/a8/e5/a8e539d5a9e2098875d9861287d04095.png"
			},
			"preview": {
				"width": 97,
				"height": 150,
				"url": "https://static1.e621.net/data/preview/a8/e5/a8e539d5a9e2098875d9861287d04095.jpg"
			},
			"sample": {
				"has": false,
				"height": 1131,
				"width": 732,
				"url": "https://static1.e621.net/data/a8/e5/a8e539d5a9e2098875d9861287d04095.png"
			},
			"score": {
				"up": 12,
				"down": -1,
				"total": 11
			},
			"tags": {
				"general": [
					"ambiguous_gender",
					"anthro",
					"armor",
					"boots",
					"city",
					"clothing",
					"covering",
					"covering_face",
					"eyewear",
					"female",
					"feral",
					"floppy_ears",
					"footwear",
					"fur",
					"glasses",
					"hair",
					"hair_over_eyes",
					"hat",
					"headgear",
					"headwear",
					"heart_pattern",
					"horn",
					"male",
					"running",
					"scarf",
					"speech_bubble",
					"star",
					"text",
					"video_games"
				],
				"species": [
					"bovid",
					"canid",
					"canine",
					"caprine",
					"darkner",
					"dinosaur",
					"fox",
					"goat",
					"green_yoshi",
					"human",
					"humanoid",
					"mammal",
					"reptile",
					"scalie",
					"yoshi"
				],
				"character": [
					"eslira_(paper_trail)",
					"kris_(deltarune)",
					"lancer_(deltarune)",
					"ralsei",
					"susie_(deltarune)"
				],
				"copyright": [
					"deltarune",
					"mario_bros",
					"nintendo",
					"sonic_the_hedgehog_(series)"
				],
				"artist": [
					"lynxgriffin"
				],
				"invalid": [],
				"lore": [],
				"meta": [
					"comic",
					"english_text"
				]
			},
			"locked_tags": [],
			"change_seq": 27127537,
			"flags": {
				"pending": false,
				"flagged": false,
				"note_locked": false,
				"status_locked": false,
				"rating_locked": false,
				"deleted": false
			},
			"rating": "s",
			"fav_count": 20,
			"sources": [
				"https://lynxgriffin.tumblr.com/post/182949981448/well-thats-some-unfortunate-timing-so-yeah",
				"https://66.media.tumblr.com/5688a544cc64f7ba2b0e5a6ef2bc23db/tumblr_pn98zpbCYQ1rz0anso5_1280.png"
			],
			"pools": [
				16238
			],
			"relationships": {
				"parent_id": null,
				"has_children": false,
				"has_active_children": false,
				"children": []
			},
			"approver_id": 169756,
			"uploader_id": 542812,
			"description": "",
			"comment_count": 3,
			"is_favorited": false,
			"has_notes": false
		}
```

so, for pools, I have only this:

```
"pools": [
    16238
]
```

No name provided. Honestly, if not for recent change in the API, even that was not there. There is an API to get a name from an id, but rules are to make one request per second, and that means either pools or posts, or a queue for different APIs, and that's just hard. The way I see it, another script should run first get all id-pools pairs and then use them. Not this week, not these month, but I think this summer at most.

**Comment by huskelord** [comment]
I tried the quick fix before but it would just flood the main folder with images that were not in pools so i'm just going without pools for now.

As for naming the pool folders after the id vs the name I undersood before i just forgot to change names to id in the example above when i edited it from a copy of one of the examples even higher up.

**Comment by huskelord** [comment]
I might have found some more bugs or it could have been an error on my end since I decided to work with pools in wrong places than no pools whatsoever

I'll need to do some more testing but i should get back to you tomorrow

**Comment by huskelord** [comment]
Ok so there is potentaly two new bugs and one needed addition to the code

Bug 1: 
Pools won't donwload and thus won't add any images that are not in the main configs so safe images don't show up since i don't call for them in the main configs only in the pool config (or they get pruned by one or both of the main configs) 

Bug2: 
Any images that are not my the same artist but in the same pool don't get added to that pool (example pool 20119)

Needed addition: 
Some pools are not in order because they have out of order id's
Would it be possible to have pooled images sorted by thier order in pool with the format line in the Default section of the pools config 

Here are a copy of my configs
[1 configs API copy.txt](https://github.com/lurkbbs/e621dl/files/4897132/1.configs.API.copy.txt)
[2 configs DB copy.txt](https://github.com/lurkbbs/e621dl/files/4897135/2.configs.DB.copy.txt)
First one is to donwload only posts of artists I have added
Second one is to copy downloaded posts into further files if im looking for a spacific thing without donwloading new posts

**Comment by lurkbbs** [comment]
Please drop `pools.ini` and `pools.generated` (should be in `configs` folder). You can zip it to a single archive and drop it here.

Unless there's something wrong with pools.ini, it's a bug.

**Comment by huskelord** [comment]
[pool stuff.zip](https://github.com/lurkbbs/e621dl/files/4899201/pool.stuff.zip)

Here is the last bit of info for my most recent run by the way don't know if it will help

status: Downloading files
checked tag: all tags are valid
current config: 2 configs DB.ini
current section: <everything>
recent warning: None so far
recent file downloaded: None so far
connection retries: None so far
posts so far: 1001430
already exist: 389284
downloaded: None so far
copied: 52732
filtered: 238477323
not found on e621: 841308
Traceback (most recent call last):
  File "e621dl.py", line 824, in <module>
  File "e621dl.py", line 317, in main
  File "e621dl.py", line 353, in process_config
  File "e621dl_lib\local.py", line 535, in get_config
  File "configparser.py", line 718, in read_file
  File "configparser.py", line 1067, in _read
configparser.DuplicateSectionError: While reading from 'configs/pools.generated' [line 5404]: section 'artist/matemi/.images/..everything/pools/20621' already exists
[530528] Failed to execute script e621dl

D:\e621.net>pause
Press any key to continue . . .

**Comment by lurkbbs** [comment]
This helps a lot!

There is a bug, after all in how this `pools.generated` is made. And luckily, it should be easily fixed. Wait a bit, less then 38 minutes.

**Comment by lurkbbs** [comment]
[e621dl.zip](https://github.com/lurkbbs/e621dl/files/4899457/e621dl.zip)

OK, honestly, I don't know how this bug even exists (my first guess was wrong), so the solution is quite speculative, but as a quick fix it may work.

**Comment by lurkbbs** [comment]
False alert!

Nope, this fix won't fix a thing

**Comment by huskelord** [comment]
Take your time. I'm in no hurry.

**Comment by lurkbbs** [comment]
[e621dl.zip](https://github.com/lurkbbs/e621dl/files/4899575/e621dl.zip)

Well I am :-D

**Comment by lurkbbs** [comment]
You know, you always give me detailed reports and always give me configs. And always thanked. And always fast. I really appreciate this. But it's 12:21 AM here, so if it's not working I'll get to it on the weekend.

Edit: Just to make sure, this is not "everything is shit before 'but'" thing. No, I really appreciate you work on making this fork of e621dl bugless. Thanks.

**Comment by huskelord** [comment]
I have never been able to figure out python or get it to work so I won't be able to test it until its a .exe

Don't worry about getting chaned over tonight i can wait

Also im glad to help and i apreciate the thanks

**Comment by jahkyah** [comment]
I can run the program as source so I can help if needed.

I will try the attached ed621dl.zip to see if I can replicate the bug!

EDIT EDIT: When I get home from work.

**Comment by huskelord** [comment]
sorry for the late reply jahlyah 
if you want to check it have the downloader version posted most recently on this issure and run my config
[test config.txt](https://github.com/lurkbbs/e621dl/files/4900376/test.config.txt) (I edit it so it shoudl be a relativly fast run) in a area away from your main downloader set up and see if any of the bugs below are fixed 
(you may need to run it twice to get the pool.ini generated) 
(if it doesn't generate delete my all my config data except the settings for the first run then replace what was deleted)

bug A) go into the main folder or into the subfolders
bug B) It doesn't donwload all the images in the pool
   problem part 1) be it safe images (if your main config filters those out)
   problem part 2) or images that do not have source tags in the main seach (in my case artist)

Since i don't know if what bug or bugs lurkbbs made the quick fix for if it worked.

Your going to want to check both folders (aeolus06 and bigdad) inside the artist folder in downloads.
 
If there is no pool folder in the main folder of either artist (which means bug A isnt fixed) then you will need to go into the .images folder, then the ..everyting folder, to find the pool folder.

go into the pool folder and find pool 20991 (bigdad is the main folder) and see if there is more than two images
if there are more than, two bug B1 is fixed.
if tehre are only two, bug B1 is not fixed

go into the pool folder and find pool 20119 (aeolus06 is the main folder) and see if there is more than one image
if there are more than, one bug B2 is fixed.
if tehre are only one, bug B2 is not fixed

tell us your results as soon  as you can

**Comment by lurkbbs** [comment]
> I have never been able to figure out python or get it to work so I won't be able to test it until its a .exe

Then here's the executable. Hope this works.

[e621dl.zip](https://github.com/lurkbbs/e621dl/files/4904523/e621dl.zip)

**Comment by huskelord** [comment]
Running it now

**Comment by huskelord** [comment]
Ok just ran it and nothing looks to be fixed at least on the bugs that I found 

These are the bugs I now about

bug A) go into the main folder or into the subfolders
bug B) It doesn't donwload all the images in the pool if they are not called for in the main config (images that are rated safe are not called for in my main config) or in the search catagories the pool folder is requested example below

IF
image A B and C are in Pool Z
Image A and B are by artist Y
Image C is by Artist X
Image B and C are rated qestionable and explicit 
Image A is rated Safe

Thus if i call for pools in my artist Y folder I only get image B since Image A is safe and Image C is by a diffrent artist (diagram of situation below)

Downloads
|
|--Artist Y (main folder)
| |
| |-- Downloaded Pools (subfolder)
| | |
| | |-- Pool Z
| | | |
| | | |-- image B
| |
| |-- .Images (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.
| |
| |-- .Animations (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.
|
|--Artist X (main folder)
| |
| |-- Downloaded Pools (subfolder)
| | |
| | |-- Pool Z
| | | |
| | | |-- image C
| | |
| | |-- etc.
| |
| |-- .Images (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.
| |
| |-- .Animations (subfolder)
| | |
| | | -- sub-catagory
| | |
| | | -- sub-catagory
| | |
| | |-- etc.

Edit1: Plus there is also the thing where some pools are in a diffrent order than the ID numbering of the images example being pool 11563 the Silver Soul Series by Matemi
(to simplfy it pool is in order the images that are in it were posted out of order thus the ID numbering messes with the pool order)
 If i could add the order of the posts via format line in the [Defalut] section so the image would be named something like this
post number 1.1572867.jpg

**Comment by lurkbbs** [comment]
> Ok just ran it and nothing looks to be fixed at least on the bugs that I found

:-(

OK, I'll try to reproduce it on my PC and fix it. Not sure how long it will take.

**Comment by huskelord** [comment]
Nope problem thanks for all your hard work.

**Comment by lurkbbs** [comment]
[e621dl.zip](https://github.com/lurkbbs/e621dl/files/4907737/e621dl.zip)

OK, this should do it (emphasis on should).

**Comment by lurkbbs** [comment]
Oh, and another thing. You pool.ini is not properly configured. Especially, there is no

```ini
make_hardlinks = true
make_cache = true
prune_downloads = true
prune_cache = true
login = <your e621 login>
api_key = <your e621 api key generated in account settings>
```

**Comment by huskelord** [comment]
... could that have been the poblem all along?

**Comment by huskelord** [comment]
Ok so i ran the new version but it keeps crashing or stoping early so i made the changes to the pool.ini an i'm going to run it again to see if it fixes anyting also i don't get any error messages or anyting like that. it just stops

**Comment by huskelord** [comment]
Ok a bunch of stuff happened so i'm going to do a full wipe and run the new program start to finish. Ill get back when its fully done
Edit: just currious what bug were you working on before exactley

**Comment by huskelord** [comment]
Ok so its still running but the post that were not called in the config itself (aka the images in the pool that would have been skipped if they weren't in the pool) are not there. 
Just to help me understand and maybe show some problems could you answer these questions

is the _pool_post_strategy =_ actually downloading images from pools or sorting images into pools

If it is downloading images, when exactly are those images in the pools that are not normaly donaloaded by the config supposed to be downloaded? 

Is the pools.generated file supposed to be a config to say download these pools or is it just a file to show what pool folders were generated

Is the pools config that gets generated after the first run of the program supposed to be outside or inside the config folder

If the answer to the first qestion is it just sorts the images into pools, 
Would  it be possibel to have a config be generated that calls for the pools themelves to be downlaoded and then have them put into the folders that are mentioned in the pools.generated file

**Comment by huskelord** [comment]
if you need any clearification let me know

**Comment by lurkbbs** [comment]
> is the pool_post_strategy = actually downloading images from pools or sorting images into pools

Nope. It just creates subfolders and then moves/copies a file if it's in a pool. `pool_download_generate = true` is what creates 
`pools.generated` config to download remaining. You can open it in a notepad, btw.

> If it is downloading images, when exactly are those images in the pools that are not normaly donaloaded by the config supposed to be downloaded?

It's not, but in pair with `pool_download_generate` it opens `pools.generated` config and downloads all the remaining files.

> Is the pools.generated file supposed to be a config to say download these pools or is it just a file to show what pool folders were generated

First.

> Is the pools config that gets generated after the first run of the program supposed to be outside or inside the config folder

Supposed to be inside, all is good there.

> Would it be possibel to have a config be generated that calls for the pools themelves to be downlaoded and then have them put into the folders that are mentioned in the pools.generated file

I'm sorry, I don't understand. As it is, it uses multiple prefilters for unique pools and downloads them in all the folders created with `pool_post_strategy`. I don't know where else they should be placed.

Edit:
> ... could that have been the poblem all along?

Nope. There was a stupid bug (they are all stupid usually) that lead to incorrect generation of `pools.generated` and e621dl to crash on that final step.

> Edit: just currious what bug were you working on before exactley

This is what I worked on all this time. It's just, I could find the reason for it and make random guesses in hope it works, because for some reason all was working on my machine. When I first make it crash same way as on yours, I found it relatively fast and (I hope) fixed it for sure.

**Comment by huskelord** [comment]
my pools config (i think you callled it pools.ini before) that gets donwloaded on the first run (before any images get downloaded) was generated outside the config folder and I haven't put it in yet. 

The pools.generated file does get made in the config folder so that should be fine

Edit1: my new run of the program is still runing by the way (on the second config which is also offine if thats a problem) 
if it doesnt work this time ill put the pools.ini in the config folder and move my second config out to see if that fixes anyting since i don't see the missing images yet (probabley because its still running normal configs)

**Comment by huskelord** [comment]
pools.generated config is now running on the program (YES)
pool.ini outside config didn't cause problems being outside the config folder

I'll be out for a while but when i come back the program should be finished 
so ill go through everything and see if any bugs remain other than the pool folders going into the subfolders

**Comment by lurkbbs** [comment]
> pools.generated config is now running on the program (YES)
> pool.ini outside config didn't cause problems being outside the config folder

Yes!

> I'll be out for a while but when i come back the program should be finished
> so ill go through everything and see if any bugs remain other than the pool folders going into the subfolders

Thanks!

Edit:

> my pools config (i think you callled it pools.ini before) that gets donwloaded on the first run (before any images get downloaded) was generated outside the config folder and I haven't put it in yet.

Yes, pool.ini should be outside of the configs folder. I guess I should have made it so `configs.generated` placed there as well, didn't think it through. But oh well, it works anyway.

> Edit1: my new run of the program is still runing by the way (on the second config which is also offine if thats a problem)

It not a problem, obviously :-)

> if it doesnt work this time ill put the pools.ini in the config folder

That wouldn't help anyway, the app would just created new one and asked to edit it.

**Comment by huskelord** [comment]
Its done and it works perfectly I'll close this issure since we know have a pools that works and functional but ill open two new issues for the formating to fix out of order pools and another for pool folders in subfolders
